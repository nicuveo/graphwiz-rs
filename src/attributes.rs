//! # DOT Attributes.
//!
//! This module exports a [&str] pub constant for [every known DOT
//! attribute](https://graphviz.org/doc/info/attrs.html), allowing the
//! user to refer to them by name rather than with a string literal.

/// DOT attribute `area`.
pub const AREA: &str = "area";

/// DOT attribute `arrowhead`.
pub const ARROWHEAD: &str = "arrowhead";

/// DOT attribute `arrowsize`.
pub const ARROWSIZE: &str = "arrowsize";

/// DOT attribute `arrowtail`.
pub const ARROWTAIL: &str = "arrowtail";

/// DOT attribute `_background`.
pub const BACKGROUND: &str = "_background";

/// DOT attribute `bb`.
pub const BB: &str = "bb";

/// DOT attribute `beautify`.
pub const BEAUTIFY: &str = "beautify";

/// DOT attribute `bgcolor`.
pub const BGCOLOR: &str = "bgcolor";

/// DOT attribute `center`.
pub const CENTER: &str = "center";

/// DOT attribute `charset`.
pub const CHARSET: &str = "charset";

/// DOT attribute `clusterrank`.
pub const CLUSTERRANK: &str = "clusterrank";

/// DOT attribute `color`.
pub const COLOR: &str = "color";

/// DOT attribute `colorscheme`.
pub const COLORSCHEME: &str = "colorscheme";

/// DOT attribute `comment`.
pub const COMMENT: &str = "comment";

/// DOT attribute `compound`.
pub const COMPOUND: &str = "compound";

/// DOT attribute `concentrate`.
pub const CONCENTRATE: &str = "concentrate";

/// DOT attribute `constraint`.
pub const CONSTRAINT: &str = "constraint";

/// DOT attribute `Damping`.
pub const DAMPING: &str = "Damping";

/// DOT attribute `decorate`.
pub const DECORATE: &str = "decorate";

/// DOT attribute `defaultdist`.
pub const DEFAULTDIST: &str = "defaultdist";

/// DOT attribute `dim`.
pub const DIM: &str = "dim";

/// DOT attribute `dimen`.
pub const DIMEN: &str = "dimen";

/// DOT attribute `dir`.
pub const DIR: &str = "dir";

/// DOT attribute `diredgeconstraints`.
pub const DIREDGECONSTRAINTS: &str = "diredgeconstraints";

/// DOT attribute `distortion`.
pub const DISTORTION: &str = "distortion";

/// DOT attribute `dpi`.
pub const DPI: &str = "dpi";

/// DOT attribute `edgeURL`.
pub const EDGEURL: &str = "edgeURL";

/// DOT attribute `edgehref`.
pub const EDGEHREF: &str = "edgehref";

/// DOT attribute `edgetarget`.
pub const EDGETARGET: &str = "edgetarget";

/// DOT attribute `edgetooltip`.
pub const EDGETOOLTIP: &str = "edgetooltip";

/// DOT attribute `epsilon`.
pub const EPSILON: &str = "epsilon";

/// DOT attribute `esep`.
pub const ESEP: &str = "esep";

/// DOT attribute `fillcolor`.
pub const FILLCOLOR: &str = "fillcolor";

/// DOT attribute `fixedsize`.
pub const FIXEDSIZE: &str = "fixedsize";

/// DOT attribute `fontcolor`.
pub const FONTCOLOR: &str = "fontcolor";

/// DOT attribute `fontname`.
pub const FONTNAME: &str = "fontname";

/// DOT attribute `fontnames`.
pub const FONTNAMES: &str = "fontnames";

/// DOT attribute `fontpath`.
pub const FONTPATH: &str = "fontpath";

/// DOT attribute `fontsize`.
pub const FONTSIZE: &str = "fontsize";

/// DOT attribute `forcelabels`.
pub const FORCELABELS: &str = "forcelabels";

/// DOT attribute `gradientangle`.
pub const GRADIENTANGLE: &str = "gradientangle";

/// DOT attribute `group`.
pub const GROUP: &str = "group";

/// DOT attribute `headURL`.
pub const HEADURL: &str = "headURL";

/// DOT attribute `head_lp`.
pub const HEAD_LP: &str = "head_lp";

/// DOT attribute `headclip`.
pub const HEADCLIP: &str = "headclip";

/// DOT attribute `headhref`.
pub const HEADHREF: &str = "headhref";

/// DOT attribute `headlabel`.
pub const HEADLABEL: &str = "headlabel";

/// DOT attribute `headport`.
pub const HEADPORT: &str = "headport";

/// DOT attribute `headtarget`.
pub const HEADTARGET: &str = "headtarget";

/// DOT attribute `headtooltip`.
pub const HEADTOOLTIP: &str = "headtooltip";

/// DOT attribute `height`.
pub const HEIGHT: &str = "height";

/// DOT attribute `href`.
pub const HREF: &str = "href";

/// DOT attribute `image`.
pub const IMAGE: &str = "image";

/// DOT attribute `imagepath`.
pub const IMAGEPATH: &str = "imagepath";

/// DOT attribute `imagepos`.
pub const IMAGEPOS: &str = "imagepos";

/// DOT attribute `imagescale`.
pub const IMAGESCALE: &str = "imagescale";

/// DOT attribute `inputscale`.
pub const INPUTSCALE: &str = "inputscale";

/// DOT attribute `cluster`.
pub const ISCCLUSTER: &str = "cluster";

/// DOT attribute `K`.
pub const K: &str = "K";

/// DOT attribute `label`.
pub const LABEL: &str = "label";

/// DOT attribute `labelURL`.
pub const LABELURL: &str = "labelURL";

/// DOT attribute `label_scheme`.
pub const LABEL_SCHEME: &str = "label_scheme";

/// DOT attribute `labelangle`.
pub const LABELANGLE: &str = "labelangle";

/// DOT attribute `labeldistance`.
pub const LABELDISTANCE: &str = "labeldistance";

/// DOT attribute `labelfloat`.
pub const LABELFLOAT: &str = "labelfloat";

/// DOT attribute `labelfontcolor`.
pub const LABELFONTCOLOR: &str = "labelfontcolor";

/// DOT attribute `labelfontname`.
pub const LABELFONTNAME: &str = "labelfontname";

/// DOT attribute `labelfontsize`.
pub const LABELFONTSIZE: &str = "labelfontsize";

/// DOT attribute `labelhref`.
pub const LABELHREF: &str = "labelhref";

/// DOT attribute `labeljust`.
pub const LABELJUST: &str = "labeljust";

/// DOT attribute `labelloc`.
pub const LABELLOC: &str = "labelloc";

/// DOT attribute `labeltarget`.
pub const LABELTARGET: &str = "labeltarget";

/// DOT attribute `labeltooltip`.
pub const LABELTOOLTIP: &str = "labeltooltip";

/// DOT attribute `landscape`.
pub const LANDSCAPE: &str = "landscape";

/// DOT attribute `layer`.
pub const LAYER: &str = "layer";

/// DOT attribute `layerlistsep`.
pub const LAYERLISTSEP: &str = "layerlistsep";

/// DOT attribute `layers`.
pub const LAYERS: &str = "layers";

/// DOT attribute `layerselect`.
pub const LAYERSELECT: &str = "layerselect";

/// DOT attribute `layersep`.
pub const LAYERSEP: &str = "layersep";

/// DOT attribute `layout`.
pub const LAYOUT: &str = "layout";

/// DOT attribute `len`.
pub const LEN: &str = "len";

/// DOT attribute `levels`.
pub const LEVELS: &str = "levels";

/// DOT attribute `levelsgap`.
pub const LEVELSGAP: &str = "levelsgap";

/// DOT attribute `lhead`.
pub const LHEAD: &str = "lhead";

/// DOT attribute `lheight`.
pub const LHEIGHT: &str = "lheight";

/// DOT attribute `linelength`.
pub const LINELENGTH: &str = "linelength";

/// DOT attribute `lp`.
pub const LP: &str = "lp";

/// DOT attribute `ltail`.
pub const LTAIL: &str = "ltail";

/// DOT attribute `lwidth`.
pub const LWIDTH: &str = "lwidth";

/// DOT attribute `margin`.
pub const MARGIN: &str = "margin";

/// DOT attribute `maxiter`.
pub const MAXITER: &str = "maxiter";

/// DOT attribute `mclimit`.
pub const MCLIMIT: &str = "mclimit";

/// DOT attribute `mindist`.
pub const MINDIST: &str = "mindist";

/// DOT attribute `minlen`.
pub const MINLEN: &str = "minlen";

/// DOT attribute `mode`.
pub const MODE: &str = "mode";

/// DOT attribute `model`.
pub const MODEL: &str = "model";

/// DOT attribute `newrank`.
pub const NEWRANK: &str = "newrank";

/// DOT attribute `nodesep`.
pub const NODESEP: &str = "nodesep";

/// DOT attribute `nojustify`.
pub const NOJUSTIFY: &str = "nojustify";

/// DOT attribute `normalize`.
pub const NORMALIZE: &str = "normalize";

/// DOT attribute `notranslate`.
pub const NOTRANSLATE: &str = "notranslate";

/// DOT attribute `nslimit`.
pub const NSLIMIT: &str = "nslimit";

/// DOT attribute `nslimit1`.
pub const NSLIMIT1: &str = "nslimit1";

/// DOT attribute `oneblock`.
pub const ONEBLOCK: &str = "oneblock";

/// DOT attribute `ordering`.
pub const ORDERING: &str = "ordering";

/// DOT attribute `orientation`.
pub const ORIENTATION: &str = "orientation";

/// DOT attribute `outputorder`.
pub const OUTPUTORDER: &str = "outputorder";

/// DOT attribute `overlap`.
pub const OVERLAP: &str = "overlap";

/// DOT attribute `overlap_scaling`.
pub const OVERLAP_SCALING: &str = "overlap_scaling";

/// DOT attribute `overlap_shrink`.
pub const OVERLAP_SHRINK: &str = "overlap_shrink";

/// DOT attribute `pack`.
pub const PACK: &str = "pack";

/// DOT attribute `packmode`.
pub const PACKMODE: &str = "packmode";

/// DOT attribute `pad`.
pub const PAD: &str = "pad";

/// DOT attribute `page`.
pub const PAGE: &str = "page";

/// DOT attribute `pagedir`.
pub const PAGEDIR: &str = "pagedir";

/// DOT attribute `pencolor`.
pub const PENCOLOR: &str = "pencolor";

/// DOT attribute `penwidth`.
pub const PENWIDTH: &str = "penwidth";

/// DOT attribute `peripheries`.
pub const PERIPHERIES: &str = "peripheries";

/// DOT attribute `pin`.
pub const PIN: &str = "pin";

/// DOT attribute `pos`.
pub const POS: &str = "pos";

/// DOT attribute `quadtree`.
pub const QUADTREE: &str = "quadtree";

/// DOT attribute `quantum`.
pub const QUANTUM: &str = "quantum";

/// DOT attribute `rank`.
pub const RANK: &str = "rank";

/// DOT attribute `rankdir`.
pub const RANKDIR: &str = "rankdir";

/// DOT attribute `ranksep`.
pub const RANKSEP: &str = "ranksep";

/// DOT attribute `ratio`.
pub const RATIO: &str = "ratio";

/// DOT attribute `rects`.
pub const RECTS: &str = "rects";

/// DOT attribute `regular`.
pub const REGULAR: &str = "regular";

/// DOT attribute `remincross`.
pub const REMINCROSS: &str = "remincross";

/// DOT attribute `repulsiveforce`.
pub const REPULSIVEFORCE: &str = "repulsiveforce";

/// DOT attribute `resolution`.
pub const RESOLUTION: &str = "resolution";

/// DOT attribute `root`.
pub const ROOT: &str = "root";

/// DOT attribute `rotate`.
pub const ROTATE: &str = "rotate";

/// DOT attribute `rotation`.
pub const ROTATION: &str = "rotation";

/// DOT attribute `samehead`.
pub const SAMEHEAD: &str = "samehead";

/// DOT attribute `sametail`.
pub const SAMETAIL: &str = "sametail";

/// DOT attribute `samplepoints`.
pub const SAMPLEPOINTS: &str = "samplepoints";

/// DOT attribute `scale`.
pub const SCALE: &str = "scale";

/// DOT attribute `searchsize`.
pub const SEARCHSIZE: &str = "searchsize";

/// DOT attribute `sep`.
pub const SEP: &str = "sep";

/// DOT attribute `shape`.
pub const SHAPE: &str = "shape";

/// DOT attribute `shapefile`.
pub const SHAPEFILE: &str = "shapefile";

/// DOT attribute `showboxes`.
pub const SHOWBOXES: &str = "showboxes";

/// DOT attribute `sides`.
pub const SIDES: &str = "sides";

/// DOT attribute `size`.
pub const SIZE: &str = "size";

/// DOT attribute `skew`.
pub const SKEW: &str = "skew";

/// DOT attribute `smoothing`.
pub const SMOOTHING: &str = "smoothing";

/// DOT attribute `sortv`.
pub const SORTV: &str = "sortv";

/// DOT attribute `splines`.
pub const SPLINES: &str = "splines";

/// DOT attribute `start`.
pub const START: &str = "start";

/// DOT attribute `style`.
pub const STYLE: &str = "style";

/// DOT attribute `stylesheet`.
pub const STYLESHEET: &str = "stylesheet";

/// DOT attribute `class`.
pub const SVGCLASS: &str = "class";

/// DOT attribute `id`.
pub const SVGID: &str = "id";

/// DOT attribute `tailURL`.
pub const TAILURL: &str = "tailURL";

/// DOT attribute `tail_lp`.
pub const TAIL_LP: &str = "tail_lp";

/// DOT attribute `tailclip`.
pub const TAILCLIP: &str = "tailclip";

/// DOT attribute `tailhref`.
pub const TAILHREF: &str = "tailhref";

/// DOT attribute `taillabel`.
pub const TAILLABEL: &str = "taillabel";

/// DOT attribute `tailport`.
pub const TAILPORT: &str = "tailport";

/// DOT attribute `tailtarget`.
pub const TAILTARGET: &str = "tailtarget";

/// DOT attribute `tailtooltip`.
pub const TAILTOOLTIP: &str = "tailtooltip";

/// DOT attribute `target`.
pub const TARGET: &str = "target";

/// DOT attribute `TBbalance`.
pub const TBBALANCE: &str = "TBbalance";

/// DOT attribute `tooltip`.
pub const TOOLTIP: &str = "tooltip";

/// DOT attribute `truecolor`.
pub const TRUECOLOR: &str = "truecolor";

/// DOT attribute `URL`.
pub const URL: &str = "URL";

/// DOT attribute `vertices`.
pub const VERTICES: &str = "vertices";

/// DOT attribute `viewport`.
pub const VIEWPORT: &str = "viewport";

/// DOT attribute `voro_margin`.
pub const VORO_MARGIN: &str = "voro_margin";

/// DOT attribute `weight`.
pub const WEIGHT: &str = "weight";

/// DOT attribute `width`.
pub const WIDTH: &str = "width";

/// DOT attribute `xdotversion`.
pub const XDOTVERSION: &str = "xdotversion";

/// DOT attribute `xlabel`.
pub const XLABEL: &str = "xlabel";

/// DOT attribute `xlp`.
pub const XLP: &str = "xlp";

/// DOT attribute `z`.
pub const Z: &str = "z";
