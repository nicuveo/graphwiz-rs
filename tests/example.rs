use goldie;
use graphwiz::attributes as attrs;
use graphwiz::*;
use std::collections::HashMap;

#[test]
fn example() {
    let mut root = Graph::new_builder();
    root.defaults_mut(Kind::Node).extend(HashMap::from([
        (attrs::FILLCOLOR, "lavender".to_string()),
        (attrs::STYLE, "filled".to_string()),
    ]));

    let a = root.new_node("a");
    let b = root.new_node("b");
    let ab = root.new_edge(a, b);
    root.attributes_mut(ab)
        .insert(attrs::STYLE, "dotted".to_string());

    let mut cluster = root.new_cluster("box");
    let c = cluster.new_node_with(
        "c",
        HashMap::from([
            (attrs::SHAPE, "circle".to_string()),
            (attrs::FILLCOLOR, "cornflowerblue".to_string()),
        ]),
    );
    cluster.build();

    root.new_edge(c, a);
    root.new_edge(c, b);
    let graph = root.build();
    goldie::assert!(render_digraph(&graph));
}
