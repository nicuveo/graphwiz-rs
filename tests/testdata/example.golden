digraph {
    node_1 [fillcolor="lavender", label="a", style="filled"]
    node_2 [fillcolor="lavender", label="b", style="filled"]
    node_1 -> node_2 [style="dotted"]
    node_5 -> node_1
    node_5 -> node_2
    subgraph cluster_4 {
        label="box"
        node_5 [fillcolor="cornflowerblue", label="c", shape="circle", style="filled"]
    }
}
